### DISCUSSION TITLE
Use control keys as triggers

### DISCUSSION THREAD

**Discussion by twwilliams** [espanso/espanso#1259] [Feature requests and Ideas]
I would like to be able to use non-printable characters (like the Ctrl key, Alt Key, Shift key, Windows key, etc. and in combinations with each other) as triggers.

I'm not sure what the implementation should look like, but maybe define the keys as {CTRL} or {ALT} or {{CTRL}} or {{ALT}}. Beyond that, whatever you think would make sense.

The only alternative I have come up with is to continue using Autohotkey in cases where I need to use control keys as triggers, but I would like to do everything in espanso.

**Comment by mrjoecastellon** [comment]
What kind of triggers do you have in mind?

**Comment by federico-terzi** [comment]
Hey @twwilliams,

Thank you for the feedback! Could you please describe a use case that could benefit from these triggers? That would help me to better understand the problem :)

Cheers :)

**Comment by avxmw** [comment]
I'm guessing he's trying to use espanso as a command launcher like pressing Ctrl+Alt+T to spawn a terminal.

**Comment by federico-terzi** [comment]
I'm wondering whether #383 might solve this use-case, given that it will be implemented relatively soon :) The only difference is that you wouldn't be able to add the control keys multiple times. For example, CTRL+ALT+T would be allowed, but CTRL+CTRL+T wouldn't. Not sure whether that would cover this use-case :)

**Comment by tobiasbueschel** [comment]
Agreed, delimiter keys that trigger an abbreviation to expand would be great! Many other text expansions have this as well, e.g., RocketTypist: https://www.witt-software.com/rockettypist/, which implements it like this:

<img width="568" alt="Screen Shot 2022-03-28 at 2 16 14 PM" src="https://user-images.githubusercontent.com/13087421/160337355-65da424e-9fec-43a5-91e8-6a513d112548.png">

**This allows me to type:**
`br` +  `⇥` which then expands to `Best regards`

**Use case I have in mind:**

```yml
  - trigger: ":br"
    replace: |
      Best,
      {{short_name}}

  - trigger: ":br2"
    replace: |
      Best regards,
      {{long_name}}
```

Currently, typing `:br` expands right away and therefore doesn't allow me to get to `:br2`.

**Comment by ClaudioGi** [comment]
I would like to use arbitrary textual shortcuts, but without any special prefix or trailer. This would ALLOW TO TYPE ANY POSSIBLE TEXTS in a text editing application without triggering replacements. The actual trigger would be typing a trailing modifier key (just typing not a simultaneous press of any kind) to make the past text input an actual trigger. For example: s((RIGHT-CTRL}} would start SciTE, s((AltGr}} trigger replacement with the text "SciTE", s((LEFT-CTRL)) will make a screenshot, s((LEFT-SHIFT)) will start screenkey, etc.  
Implementing this feature will make Espanso being superior compared to AutoKey with a choice of a triggering keys limited to SPACE, ENTER and TAB.

**Comment by rorygamble** [comment]
I used to have AutoKey configured so that [ctrl]+[meta]+[alt] + D would insert today's date. Having a trigger like `:date` etc is too easily triggered when programming. This feature would be really useful for me, particularly since AutoKey does not support Wayland, which is now default in Ubuntu (using X11 is not a viable workaround for me because it causes other problems).

**Comment by DoctorSubtilis** [comment]
As autokey user, now interesting in espanso, I wish have a shortcut as trigger of some expansions, as I have explained [here](https://www.reddit.com/r/medlabprofessionals/comments/1b4j123/how_set_a_shortcut/).

**Comment by smeech** [comment]
#1933
